app/models/tag.rb

```diff
@@ -36,6 +36,22 @@ class Tag < ActsAsTaggableOn::Tag
     searchableAttributes %w[name short_summary]
   end
 
+  def index_to_elasticsearch
+    Search::TagEsIndexWorker.perform_async(id)
+  end
+
+  def index_to_elasticsearch_inline
+    Search::Tag.index(id, serialized_search_hash)
+  end
+
+  def serialized_search_hash
+    Search::TagSerializer.new(self).serializable_hash.dig(:data, :attributes)
+  end
+
+  def elasticsearch_doc
+    Search::Tag.get(id)
+  end
```

Comment:
Can we have comments? The previous two methods are self explanatory but these two seem to require domain language and/or information related to Elasticsearch. Also, should they be private? Or even better, should we have a sidecar module in search that gets included here?

For example:

```ruby
module TagSearchProperties # have no idea if these are properties
  after_commit :index_to_elasticsearch

  def index_to_elasticsearch
  end

  def serialized_search_hash
  end

  # ...
end

class Tag
  include serialized_search_hash
end
```

this way we start grouping code instead of enlarging the size of model classes. I think it's important to make these decisions now that we have new code, instead of deferring them for later, when we'll probably won't change them anyway.

cc @citizen428 which might have thoughts on this

Reply:
> Also, should they be private?

I would say no, in my experience these methods will be used a lot as helpers for fetching model related documents. They also make debugging super easy. 

> should we have a sidecar module in search that gets included here?

I would also say no to this as well. This is it for the methods that will be in the model. The rest will be added to the corresponding `Search::Tag` class. The reason I know this is from implementing search previously with a Rails codebase and knowing that from here on out all of the other searching logic will be VERY ES specific and therefore will not be anywhere near the model class. 

> I think we should have comments here... 

What do you suggest? I think I am too close to this to understand what might be confusing. The method name `elasticsearch_doc` tells me I am returning a document and seeing the get method makes sense since that is how you usually return information via HTTP.

Reply:
I think my question stems from the fact that as a total newbie of Elasticsearch that hasn't had the time to catch up with it yet, I do not know what calling `serialized_search_hash` and `elasticsearch_doc` should do :) Hence my request for comments. 

As you say, some of this code it's going to be very ES related, adding a line or two above the method to explain what it does it'd be awesome :D

Reply:
> my question stems from the fact that as a total newbie of Elasticsearch

I understand that people might not be familiar with Elasticsearch but I don't think it is the code's duty, in this case, to educate them on it. I think code comments should be used to explain things in the code that are unclear or obscure. I think learning about a specific database and how it works should be something that someone does on their own when they see the word Elasticsearch. The question also then becomes, do we comment every get request we make to Elasticsearch and all of our search methods? I view these methods the same as the requests we make to Redis and Postgres, we write them clearly so its obvious what database we are talking to and how we are making the request but we leave the rest of the understanding up to the user.

Reply:
@rhymes @mstruve Could we include a thorough page in our docs about how to think about/approach Elasticsearch and give short overview of its basic concepts (and maybe link to good related resources)? That might be a better home for this and help ES newbies without having to add this context in the form of code comments. I think this would be especially helpful to have somewhere in docs.dev.to since we're leaning into ES.

Reply:
@mstruve I agree in theory but we're also an open source platform with lots of contributors with all kinds of experience and as a general principle we should try to keep the code as readable and accessible as possible. I definitely agree that the code shouldn't explain ES but here the problem might simply lie in better naming. If, I'm guessing here, `elasticsearch_doc` transforms the object in a ES document, why not calling it "transform_into_elasticsearch_document`, or (following Rails convention), `as_elasticsearch_document`/`to_elasticsearch_document`, similar to the builtin `as_json`/`to_json` methods?

@vaidehijoshi good idea!

Reply:
I'm good with `as_elasticsearch_document`

Reply:
`as_elasticsearch_document` isn't bad but `as_json`/`to_json` imply a transformation of the data we have, this is a bit different bc we are actually pulling the data out of a database as is rather than transforming it in some way, hence the get request when the id argument.

Reply:
your comment kinda proves my point in away, as I didn't understand what the method was doing :P

Reply:
what about something like `retrieve_document_from_elasticsearch` ? It can't get any clearer than this :D

Reply:
What about this bc I also hate long method names, especially ones I know I will be typing a lot. Lets keep the Tag model method name as is and then change the Search::Tag method to `find_document`. THEN when you go into the search document class you see we use a get method to find the document 
```ruby
  def elasticsearch_doc
    Search::Tag.find_document(id)
  end
```
In Search::Tag
```ruby
def find_document(tag_id)
  SearchClient.get(id: tag_id, index: INDEX_ALIAS)
end
```

Reply:
Card to follow up with docs at a later point: https://github.com/thepracticaldev/dev.to/projects/6#card-32912217

Reply:
hahaha I don't have a problem with long method names :D and I feel a bit like I'm nitpicking here.

If I'm looking at the place where we use `elasticsearch_doc` (which could be far from where it's defined in the model) I still need to open the model file and then go find `Search::Tag` instead of having a method name like `find_from_search_index` (as we said, the word `elasticsearch` is going to disappear at some point).

I know we can rename methods later on but I'd rather get them right as soon as possible as experiments tend to get cementified :D 

Another possibility is to prefix all search related methods with `search_` to make it easier to spot them? Just a thought, not really convinced about this

Reply:
Given we have put all the search methods and code in a `Search` module I don't think there is also a need to prefix all of the related methods as well. When I see `find_from_search_index` I wonder then why dont we have `find_from_postgres`, `find_from_redis`. I think keeping code contained in a Search module where its a one stop shop to find what you are looking for means you don't have to have it explained everywhere throughout the code base.

Reply:
> When I see find_from_search_index I wonder then why dont we have find_from_postgres, find_from_redis.

Ah ah good point, but mainly because PostgreSQL is hidden from the ORM interface and we're not using Redis directly (only through the cache interface and the sidekiq interface). It's all about that encapsulation/interface/facade I guess :D